### DISCUSSION TITLE
Uno-Platform: Automation ids don't show on Android as they do on UWP

### DISCUSSION THREAD

**Discussion by Xiphereal** [unoplatform/uno#8125] [Q&A] [answered]
### Current behavior

I'm trying to automate UI tests via Appium on a Uno Platform app, both on Android and UWP heads.

On UWP, the custom automation ids setted for several controls work as intended, but for Android, they do not show at all. I've read the associated documentation (https://platform.uno/docs/articles/features/working-with-accessibility.html#automationid) and a blog post on this matter (https://platform.uno/blog/an-introduction-to-unit-testing-for-uno-platform-applications/, the "Enable UIAutomationMapping" section), but setting that flag both on the Android csproj and on the Shared project App.xaml.cs constructor don't have any effect.

Am I missing something? How can I use these automation ids for automating the Android head?

Thanks in advance :)

### Expected behavior

The automation ids shows on Android as they do on UWP when `Uno.UI.FrameworkElementHelper.IsUiAutomationMappingEnabled` is set to true.

**Comment by jeromelaban** [comment, marked as answer]
UI Testing is generally supported when using Uno.UITest. You may find [additional information here](https://github.com/unoplatform/Uno.uitest) that can be used when working with appium.

**Reply by Xiphereal** [reply]
I've already checked the info on the Uno.UITest repo, but couldn't find anything new about this issue. Any concrete information you wanted to point me to?

**Reply by jeromelaban** [reply]
This block, specifically:
```
 <PropertyGroup>
 	<IsUiAutomationMappingEnabled>True</IsUiAutomationMappingEnabled>
 </PropertyGroup>
 ```

**Reply by Xiphereal** [reply]
Sorry if I wasn't clear about that, but I've already setted that variable, both on the Android head csproj and on the Shared project App.xaml.cs constructor, and it doesn't seem to have any effect.

**Reply by jeromelaban** [reply]
Thanks. Can you provide an application sample that you've setup and show which controls you expect to be visible in you automation tool?

**Reply by jeromelaban** [reply]
Thanks for the repro.

You need to set `AutomationProperties.AccessibilityView="Raw"` on the elements.

With more details, here's what you get when using `<IsUiAutomationMappingEnabled>True</IsUiAutomationMappingEnabled>`:

![image](https://user-images.githubusercontent.com/5839577/155027490-2b7cd7e6-aae6-4e3c-b985-2be9f1306444.png)

The way android works for marking items for automation is through the `ContentDescription` property:

https://github.com/unoplatform/uno/blob/4763e654dc3c383eb77036a10089aa4681357747/src/Uno.UI/UI/Xaml/Automation/AutomationProperties.cs#L142-L145

which is generally overridden for `ContentControl` based controls (`Button`, `CheckBox`, `RadioButton`, etc...) with the text contents of the AutomationPeer, when initialized here:
https://github.com/unoplatform/uno/blob/4763e654dc3c383eb77036a10089aa4681357747/src/Uno.UI/UI/Xaml/Automation/Peers/FrameworkElementAutomationPeer.Android.cs#L37-L42
and populated here:
https://github.com/unoplatform/uno/blob/a16a92f29e4c4f8f7d04fa63023763047d897602/src/Uno.UI/UI/Xaml/Controls/ContentControl/ContentControl.cs#L499-L512

You can see for instance that the `Button`'s content is `Button`, and if you want to have `UnoButton` instead, you'll need to set `AutomationProperties.AccessibilityView="Raw"` as well on the button.

**Reply by Xiphereal** [reply]
Yeah, that was it, thanks :D. I think it would be useful to include this on the docs, so no one gets confused in the future :)

**Reply by jeromelaban** [reply]
Indeed, https://github.com/unoplatform/Uno.UITest/pull/73 and https://github.com/unoplatform/uno/pull/8164 will help.

**Comment by Suriman** [comment]
Hi, I’m part of his team and speak on his behalf. Here is the application sample
[UnoApp.zip](https://github.com/unoplatform/uno/files/8110314/UnoApp.zip)
:
